betty/logging.py

```diff
@@ -0,0 +1,97 @@
+"""
+Logging utilities.
+"""
+
+import contextlib
+import logging
+import threading
+from asyncio import (
+    CancelledError,
+    get_running_loop,
+    run_coroutine_threadsafe,
+    to_thread,
+)
+from collections.abc import Awaitable, Callable, Mapping
+from functools import partial
+from queue import Empty, Queue
+from time import sleep
+from types import TracebackType
+from typing import Self, final
+
+from typing_extensions import override
+
+from betty.functools import Result, ResultUnavailable, suppress
+from betty.locale.localizable import plain
+from betty.user import User, Verbosity
+
+VERBOSITY_TO_LOG_LEVEL: Mapping[Verbosity, int] = {
+    Verbosity.ESSENTIAL: logging.ERROR,
+    Verbosity.VERBOSE: logging.INFO,
+    Verbosity.MORE_VERBOSE: logging.ERROR,
+    Verbosity.MOST_VERBOSE: logging.NOTSET,
+}
+
+
+@final
+class UserHandler(logging.Handler):
+    """
+    Output log records through a :py:class`betty.user.User`.
+    """
+
+    def __init__(self, user: User):
+        super().__init__(-1)
+        self._user = user
+        self._result = Result(self._consume)
+        self._thread = threading.Thread(
+            name=self.__class__.__name__, target=suppress(self._result, BaseException)
+        )
+        self._queue = Queue[Callable[[], Awaitable[None]]]()
+        self._finish = threading.Event()
+        self._loop = get_running_loop()
+
+    async def __aenter__(self) -> Self:
+        self._thread.start()
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        self._finish.set()
```

Comment:
We probably want to communicate cancellations directly and quit the thread immediately, because at that point the entire application is unstable.